src/common/threadpool.h

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2017, The Monero Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <boost/thread/condition_variable.hpp>
+#include <boost/thread/mutex.hpp>
+#include <boost/thread/thread.hpp>
+#include <cstddef>
+#include <functional>
+#include <utility>
+#include <vector>
+
+namespace tools
+{
+//! A global thread pool
+class threadpool
+{
+public:
+  static threadpool& getInstance() {
+    static threadpool instance;
+    return instance;
+  }
+
+  // The waitobj lets the caller know when all of its
+  // tasks are completed.
+  typedef struct waitobj {
+    boost::mutex mt;
+    boost::condition_variable cv;
+    int num;
+    waitobj() : num(0){}
+  } waitobj;
+
+  // Submit a task to the pool. The obj pointer may be
+  // NULL if the caller doesn't care to wait for the
+  // task to finish.
+  void submit(waitobj *obj, std::function<void()> f);
+
+  // Wait for a set of tasks to finish.
+  void wait(waitobj *obj);
+
+  private:
+    threadpool();
+    ~threadpool();
+    typedef struct entry {
+      waitobj *wo;
+      std::function<void()> f;
+    } entry;
+    std::deque<entry> queue;
+    boost::condition_variable has_work;
+    boost::mutex mutex;
+    std::vector<boost::thread *> threads;
```

Comment:
`boost::thread` is movable so it does not have to be a pointer here. Saves a tiny bit of headache.

Reply:
I didn't see much difference, can you explain? With the explicit new in the constructor I have a corresponding delete in the destructor - would any of that be simpler by removing the pointer?

Reply:
If the `threadpool` constructor throws, its destructor never executes. The two obvious throwing functions (but still unlikely) are the `threads.push_back` and the `boost::thread` constructor. So there is _technically_ a memory leak using the raw pointer. The system should shutdown in that situation so its probably not ever going to be much of a problem.

The change reduces the line count by 1, so it seemed easy enough.